### DISCUSSION TITLE
tutorials/math/vector_math

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#265] [User-contributed notes]
# tutorials/math/vector_math

Introduction: This tutorial is a short and practical introduction to linear algebra as it applies to game development. Linear algebra is the study of vectors and their uses. Vectors have many appli...

https://docs.godotengine.org/en/stable/tutorials/math/vector_math.html

**Comment by Maffey** [comment]
Glad you also tackle topics that are not directly connected with Godot itself. Thanks!

**Comment by PDJEY** [comment]
I'm in college studying stuff like this. This small doc was able to help me understand what college professors couldn't in an entire semester

**Comment by mixx99** [comment]
If you are looking to build a strong foundation in linear algebra, the Gilbert Strang lectures on Linear Algebra (MIT) are an excellent place to start — they are available on YouTube. His book, Introduction to Linear Algebra, is also highly recommended.

**Comment by smartcode444** [comment]
Wow, I easily understood vector math through this article. Thanks godot team

**Comment by volstas** [comment]
This was more confusing than it needed to be because of interchanging terminology.

E.g. sometimes it's called magnitude sometimes it's called length.

Same with unit vectors. Introduced as unit vectors. Says it's sometimes called direction vectors or normals. But then the doc proceeds to call it "normals" the entire time.

I think it would be clearer if it was introduced as the thing it's called in code. And then mention the math terms just as a side, but primarily terminology should be the code terminology.

**Comment by Printographe** [comment]
One important corrolary of one of the formulas that you've shown on the dot product is that the square root of the dot product of a vector with itself is the length of that vector. It can be shown using both formulas : 
 
Reminder of the distance formula of a vector v in 2D :  distance(v) = sqrt( pow(v.x, 2) + pow(v.y, 2) ) 

Now back to the dot product formula, if we plug the vector a as both arguments : 
dot(a, a) = a.x * a.x + a.y * a.y = pow(a.x, 2)+ pow(a.y, 2). 

Taking the square root of dot(a, a) we find our well known formula.

Using the other formula, we just have to keep in mind that the cosine of two vectors asks the question:
 "from -100% to 100% how a similar are these two vectors", intuitively the cosine of the same vector will yield 100%.

now back to the formula : 
dot(a, a) = ||a|| ||a|| cos(a, a) = distance(a) * distance(a) * 1 = pow( distance(a), 2 ) * cos(a, a) = pow(distance(a), 2)
Again, taking the square root will remove that power of 2.

This can be used for instance for when you only have the velocity but you don't know at what speed you're currently going, although I wouldn't recommend to write var speed = sqrt(self.velocity.dot(self.velocity)) (at least not with a comment) because it's not readable. But this should give a glance on how fundamental some of these formulas are :)

**Comment by pkq403** [comment]
really good article, thanks

**Comment by screen-pixl-4096** [comment]
Stopped understanding it during the whole bounce function thing but maybe I can come back some day and understand this
